src/Node/TypesNode.php

```diff
@@ -0,0 +1,53 @@
+<?php
+
+namespace Twig\Node;
+
+use Twig\Attribute\YieldReady;
+use Twig\Compiler;
+use Twig\Node\Expression\ArrayExpression;
+use Twig\Node\Expression\ConstantExpression;
+use Twig\Node\Expression\NameExpression;
+
+/**
+ * Represents a types node.
+ *
+ * @author Jeroen Versteeg <jeroen@alisqi.com>
+ * @see    https://github.com/twigphp/Twig/issues/4165
+ */
+#[YieldReady]
+class TypesNode extends Node implements NodeCaptureInterface
+{
+    public function __construct(ArrayExpression $typesNode, int $lineno)
+    {
+        $this->validateMapping($typesNode);
+
+        parent::__construct(['mapping' => $typesNode], [], $lineno);
+    }
+
+    protected function validateMapping(ArrayExpression $typesNode): void
```

Comment:
this validation looks useless as the TypesTokenParser already guarantees that this is respected.

Reply:
But the constructor takes a (generic) `ArrayExpression`.

I considered creating a specialized `NameToStringMappingExpression` to lock this down, but thought that would be overkill. What would you advise?

I _do_ want to nail down the structure of the expression to prevent errors down the line (post-construction). It doesn't hurt, right?

Reply:
@ruudk suggested taking in an `array` in the constructor and store it as an attribute instead.

Reply:
I also think that this does not need to be an ArrayExpression (especially as we want to be a lot more strict)

Reply:
What if we make it an array typed as `array<string, string>`. Then store it in the attributes of the TypesNode? 

Optional variables will have an `?` at the end of the string key.

Reply:
> array<string, array{type: string, optional: bool}>

Also fine!

Reply:
Done!